mat/matrix.go

```diff
@@ -759,55 +763,19 @@ func Min(a Matrix) float64 {
 func Norm(a Matrix, norm float64) float64 {
 	r, c := a.Dims()
 	if r == 0 || c == 0 {
-		panic(ErrShape)
+		panic(ErrZeroLength)
```

Comment:
Trace on an empty matrix returns zero, so I think we should either remove the panic in Norm or add it to Trace.

Reply:
SGTM. Which would you prefer? I'm happy either way.

Reply:
This is almost a philosophical topic similar to empty set.

We can view an empty matrix as a zero matrix of any size; then Trace, Norm, etc would naturally return zero without panic but that zero would be different from those operations on non-empty matrices. However, does this distinction matter? This would be similar as returning zero by definition/convention.

We can also view it as an empty container and start from there. Trace is a sum and a sum of an empty sequence is by convention 0. Frobenius norm is also a sum. But 1-norm and inf-norm are _maximum_ over columns/rows and maximum over an empty set would be ... minus infinity? NaN? That seems worse than a panic. And panicking only for certain norms is too much hassle for a minor corner case.

Or we can panic always. So we're where we started :-)

I think I'm slightly leaning toward the first option - return 0?

Reply:
I'm OK with returning 0 for the trace. If we do that we should probably return the supremum for 1-norm and inf-norm since it seems like the most useful definition; if returning a value from an empty matrix is seen as OK.

Reply:
It seems to me that you'd prefer a panic for empty matrices and I agree that it's much easier to consistently do it (also over time) in contrast to returning values like -inf from Norm (or mat.Max). Should we then settle on: operations on empty matrices panic unless used as a destination/receiver?